mediacloud/mediawords/tm/test_fetch_topic_tweets.py

```diff
@@ -0,0 +1,297 @@
+        got_date = datetime.datetime.strptime(topic_tweet['publish_date'], '%Y-%m-%d 00:00:00')
+        assert got_date == expected_date
+
+        assert topic_tweet['content'] == tweet_data['tweet']['text']
+
+
+def validate_topic_tweet_urls(db: DatabaseHandler, topic: dict) -> None:
+    """Validate that topic_tweet_urls match what's in the tweet JSON data as saved in topic_tweets."""
+    topic_tweets = db.query(
+        """
+        select *
+            from topic_tweets tt
+                join topic_tweet_days ttd using (topic_tweet_days_id)
+            where
+                ttd.topics_id = ?
+        """,
+        topic['topics_id']).hashes()
+
+    expected_num_urls = 0
+    for topic_tweet in topic_tweets:
+        data = mediawords.util.json.decode_json(topic_tweet['data'])
+        expected_num_urls += len(data['tweet']['entities']['urls'])
+
+    # first sanity check to make sure we got some urls
+    num_urls = db.query("select count(*) from topic_tweet_urls").flat()[0]
+    assert num_urls == expected_num_urls
+
+    total_json_urls = 0
+    for topic_tweet in topic_tweets:
+
+        ch_post = mediawords.util.json.decode_json(topic_tweet['data'])
+        expected_urls = [x['expanded_url'] for x in ch_post['tweet']['entities']['urls']]
+        total_json_urls += len(expected_urls)
+
+        for expected_url in expected_urls:
+            got_url = db.query("select * from topic_tweet_urls where url = $1", expected_url).hash()
+            assert got_url is not None
+
+    assert total_json_urls == num_urls
+
+
+def run_fetch_topic_tweets_test(db: DatabaseHandler) -> None:
+    """Run fetch_topic_tweet tests with test database."""
+    topic = mediawords.test.db.create_test_topic(db, 'test')
+
+    test_dates = get_test_date_range()
+    topic['start_date'] = test_dates[0]
+    topic['end_date'] = test_dates[1]
+    topic['ch_monitor_id'] = 123456
+    db.update_by_id('topics', topic['topics_id'], topic)
+
+    mediawords.tm.fetch_topic_tweets.fetch_topic_tweets(db, topic['topics_id'], MockTwitter, MockCrimsonHexagon)
+
+    topic_tweet_days = db.query("select * from topic_tweet_days").hashes()
+    assert len(topic_tweet_days) == LOCAL_DATE_RANGE + 1
+
+    start_date = datetime.datetime.strptime(topic['start_date'], '%Y-%m-%d')
+    test_days = [start_date + datetime.timedelta(days=x) for x in range(0, LOCAL_DATE_RANGE)]
+    for d in test_days:
+        topic_tweet_day = db.query(
+            "select * from topic_tweet_days where topics_id = $1 and day = $2",
+            topic['topics_id'], d
+        ).hash()
+        assert topic_tweet_day is not None
+
+        validate_topic_tweets(db, topic_tweet_day)
+
+    validate_topic_tweet_urls(db, topic)
+
+
+def test_fetch_topic_tweets() -> None:
+    """Generate test database and run tests."""
+    mediawords.test.db.test_on_test_database(run_fetch_topic_tweets_test)
+
+
+@unittest.skipUnless(os.environ.get('MC_REMOTE_TESTS', False), "remote tests")
```

Comment:
What's this `MC_REMOTE_TESTS` environment variable, where is it being defined?

I'm afraid that this test will end up never being run because no one will ever remember to set `MC_REMOTE_TESTS`. To get the tests run against some sort of a remote API (e.g. Amazon S3), so far we've been doing this:

1. Come up with a one or more environment variables that will hold credentials needed for the remote test to run, e.g. `MC_CRIMSON_HEXAGON_TEST_KEY=0123456789abcdef`.
2. Encrypt them using [Travis encryption](https://docs.travis-ci.com/user/encryption-keys/#Usage) and put the encrypted values at the top of `.travis.yml`.
3. In the test, if the environment variable with test credentials is set, fetch the configuration `dict`, set the credentials from the environment variable, `set_config(updated_configuration)`, run the test, and lastly set it back to the original configuration.

Example with Bit.ly (not yet merged into `master`): https://github.com/berkmancenter/mediacloud/blob/bitly_python/lib/MediaWords/Test/Bitly.pm#L286-L348. The method is also being used with Amazon S3, Superglue, Univision, and Facebook API.

I'm not sure if our current method is the optimal one to test against remote APIs (and @ColCarroll would probably say that we shouldn't be hitting remote APIs at all because principle > necessity), but I'd argue that the current approach is better than having someone to set an arbitrary environment variable locally and update `mediawords.yml` manually with a bunch of API credentials - the "remote" tests get run on Travis on every build at least.